newsfragments/4607.changed.md

```diff
@@ -0,0 +1,2 @@
+- `IntoPyCallbackOutput` switched blanket implementation to `IntoPyObject`
+- `IntoPyCallbackOutput` gained `'py` lifetime parameter
```

Comment:
`IntoPyCallbackOutput` is hidden, so I don't think this needs a changelog..unless it enables users to do things they couldn't do before (does it?). In which case we should document that instead.

Reply:
Oh, I didn't notice that the whole module was hidden (maybe it should be moved into `impl_`). This should not enable any new things. The trait is used however as a public bound in `PyModuleMethods::add_wrapped` and `PyCFunction` constructors. Because the blanket impl changed there could be breakage for downstream types that do not (yet) implement `IntoPyObject`. The additional lifetime is less of a concern in this case I guess.

Reply:
I think I'd suggest the following:
- Just skip the CHANGELOG from this PR
- Two follow ups (ideally to complete before 0.23), let's open them as issues:
  - Let's move `pyo3::callback` module into the `impl_` submodule
  - Let's move the function constructors to a new trait. They don't need all the generic complexity of `IntoPyCallbackOutput`, it would be probably good enough to have a simple trait which is implemented for both `T: IntoPyObject` and `Result<T, E> where T: IntoPyObject, E: Into<PyErr>`